File: macos/Runner/MethodChannels/NativePathChannel.swift
import FlutterMacOS
import Foundation
import OSLog

final class NativePathChannel {
    static let channelName = "ch.cbue.protonContactBridge/native_path"
    private static let logger = Logger(
        subsystem: "ch.cbue.protonContactBridge",
        category: "NativePathChannel"
    )

    private let channel: FlutterMethodChannel

    init(with messenger: any FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: messenger
        )
        channel.setMethodCallHandler { [weak self] call, result in
            guard let self else {
                result(
                    FlutterError(
                        code: "channel_unavailable",
                        message: "Native Path channel was released",
                        details: nil
                    )
                )
                return
            }

            self.handle(call: call, result: result)
        }
    }

    deinit {
        channel.setMethodCallHandler(nil)
    }

    private func handle(
        call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        Self.logger.notice(
            "received method call: \(call.method, privacy: .public)"
        )
        switch call.method {
        case "databasePath":
            guard let databaseURL = try? Database.url() else {
                result(FlutterError(
                    code: "container_unavailable",
                    message: "Unable to access shared app group container",
                    details: nil
                ))
                return
            }
            result(databaseURL.path)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/app/src/main/kotlin/ch/cbue/proton_contact_bridge/method_channels/NativePathChannel.kt
package ch.cbue.proton_contact_bridge.method_channels

import android.content.Context
import android.util.Log
import ch.cbue.proton_contact_bridge.database.ContactsDatabase
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class NativePathChannel(private val context: Context) : AppMethodChannel {
    companion object {
        private const val TAG = "NativePathChannel"
    }

    override val channelName = "ch.cbue.protonContactBridge/native_path"

    private var channel: MethodChannel? = null

    override fun register(messenger: BinaryMessenger) {
        channel = MethodChannel(messenger, channelName)
        channel?.setMethodCallHandler { call, result ->
            handle(call, result)
        }
    }

    override fun tearDown() {
        channel?.setMethodCallHandler(null)
        channel = null
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        Log.i(TAG, "received method call: ${call.method}")

        when (call.method) {
            "databasePath" -> {
                try {
                    val dbFile = ContactsDatabase.path(context)
                    result.success(dbFile)
                } catch (e: Exception) {
                    result.error(
                        "container_unavailable",
                        "Unable to access safe app storage path",
                        e.localizedMessage
                    )
                }
            }

            else -> result.notImplemented()
        }
    }
}